career-mentorship-program/2021-spring/matching_code.py

```diff
@@ -410,9 +503,66 @@ def compute_weight_and_statistics(self):
         if year_difference == 1 or year_difference == 2:
             weight += MENTOR_CLOSE_IN_YEAR_BONUS
 
-
         """ !!! Add any new matching criteria here !!! """
 
+        # 4. Hobbies
+
+        num_hobbies_overlap = len(self.mentee.hobbies & self.mentor.hobbies)
+        num_mentee_hobbies = len(self.mentee.hobbies)
+        hobbies_overlap_fraction = num_hobbies_overlap / num_mentee_hobbies
+        statistics["hobbies_overlap_fraction"] = hobbies_overlap_fraction
+        weight += HOBBIES_TOPICS_MULTIPLIER * hobbies_overlap_fraction
+
+        # function for gender and race
+        def gender_and_race(mentor_preference, mentee_preference, stat_field1, stat_field2, bonus):
```

Comment:
Similar comment as above, I think it's best to not have side effects like changing the weight and the statistics fields hidden in this function but to instead have them in the top level function. In this case, doing the gender and race calculations require accessing so many different fields on the mentor and mentee classes that it's hard to abstract any of the logic away into a function. I would just keep all the logic at the top level like this:

```
both_prefer_gender = self.mentor.does_prefer_gender and self.mentee.does_prefer_gender
same_gender = self.mentor.gender == self.mentee.gender
# add statistics here
if both_prefer_gender and same_gender:
    self.weight += GENDER_SAME_BONUS

both_prefer_race = self.mentor.does_prefer_race and self.mentee.does_prefer_race
race_overlap_fraction = compute_fraction_overlap(self.mentee.race, self.mentor.race)
# add statistics here
if both_prefer_race and same_race:
    self.weight += RACE_SAME_BONUS * race_overlap_fraction
```